stdlib/asm/crypto/fri/frif2e2.masm

```diff
@@ -0,0 +1,288 @@
+    movup.7                     #[?,poe,poe,V,p%d/2,d/2,p,t_d,..]
+    exec.next_pos_exp           #[poe_sq,xs,V,p%d/2,d/2,p,t_d,..]
+    swap.8 drop                 #[xs,V,p%d/2,d/2,poe_sq,t_d,..]
+    push.0                      #[0,xs,V,p%d/2,d/2,poe_sq,t_d,a1,a0,..]
+    movup.11 movup.11           #[a1,a0,0,xs,v3,v2,v1,v0,p%d/2,d/2,poe_sq,t_d,..]   
+    swapw                       #[v3,v2,v1,v0,a1,a0,0,xs,p%d/2,d/2,poe_sq,t_d,..] #Should simplify fri since d1 is always equal to zero
+    movup.3 movup.3
+    exec.fold_2                 #[e1,e0,p%d/2,d/2,poe_sq,t_d,..]
+
+    # Decrement the depth i.e. move one layer deeper in the folding step of FRI verification.
+    movup.5 u32wrapping_sub.1   #[t_d-1,e1,e0,p%d/2,d/2,poe_sq,add',..]
+    
+
+end
+
+#! Verify a single query. The expected stack configuration is [add,num_q,d,g,t_d,add',...]
+#! where add is the address storing the current (num_q,d,g,t_d), num_q is the index of the current query,
+#! d is the domain size of the LDE, g is its generator and add' is the address of the commitment to 
+#! the first layer with the commitments to subsequent layers laid out, in alternating order with the layer
+#! alphas, in the subsequent addresses i.e. add' - 2, add' - 1 ...
+#! Input:  #[d,p,C,t_d,e1,e0,poe,a1,a0,add_f - 2,add_f,add_i,...]
+#! Output: #[p%d/2,d/2,poe_sq,add_i + 1,add_f,add_i,rem1,rem0...]
+export.verify_query
+    exec.verify_query_layer
+    #[t_d-1,e1,e0,p%d/2,d/2,poe_sq,add',..]
+
+    dup u32checked_neq.3
+    while.true
+        exec.prepare_next
+        exec.verify_query_layer
+        dup u32checked_neq.3
+    end
+
+    ## Verify remainder 
+    drop
+    dup.8 assert_eq
+    dup.8 assert_eq
+end
+
+#! Verify that the hash of the remainder codeword is equal to the commitment provided by the prover
+#! The following implementation relies on the assumption that the blowup factor is 8 and that the degree
+#! of the remainder is zero.
+#! Input: #[p%d/2,d/2,poe_sq,add_i + 1,add_f,add_i,rem1,rem0...]
+#! Output: ()
+export.verify_remainder_com
+    dropw                  #[r1,r0,..]
+    drop drop
+    dup.1 dup.1
+    push.4.0.0.0
+    swapw
+    push.0.0.0.0
+
+    # compute the leaf
+    rpperm
+    dropw swapw dropw
+
+    # compute the Merkle tree root
+    dupw rphash     
+    dupw rphash
+    push.0.0.0.0
+    adv_loadw
+    eqw
+    assert
+    dropw dropw
+end
+
+#! This procedure verifies a FRI proof with number of queries num_q, initial domain size d, generator
+#! for this domain g, and tree depth of the initial layer Merkle tree commitment. add_i points to
+#! the address containing values common to all queries as well as to keep track of the queries
+#! that remain to be verified. add_f points to the address that will contain the result of adv_pipe.
+#! add_f - 2 * i - 2 points to the layer folding challenges alpha and add_f - 2 * i - 1 points to
+#! the i-th layer commitment.
+#! Input: #[add_i,num_q,d,g,t_d,add_f,...]
+#! Output: ()
```

Comment:
- I don't think we need to keep track of 'add_f` since I don't think we need to keep the decommitments in memory after they've been used (but I might be wrong).
- In `add_f - 2 * i - 2 points to the layer folding challenges alpha and add_f - 2 * i - 1 points to
#! the i-th layer commitment` is `add_f` supposed to be `add_i`?

Minor nits about naming/format etc.:

- I would probably use slightly more explicit names to reduce cognitive load on the reader. For example, `d -> d_size`, `t -> t_depth` (keeping `g` as is I think is fine).
- For memory references, instead of `add` prefix, I would use `ptr` suffix (this is a more commonly used convention). So, instead of `add_i`, I would say `i_ptr`.
- Also, I would probably use `layer_ptr` instead of `i_ptr` as it is a bit more descriptive about what the variable points to.
- As mentioned in one of previous comments, I'd add spaces after commas.

Reply:
Just to be clear, `f_ptr` is pointing to the commitment (and alpha) for the the initial commitment. The commitment to the next layer is in `f_ptr - 2`.
`i_ptr` is pointing to the the memory containing the tuple `(num_q, d_size, g, t_depth)` which is used across all queries and where `num_q` is decremented when a query is processed so as to keep track of of the queries remaining.
So, in the way we have it now, both these are required but I think that there is some pointers which are not necessary after I have included your suggestion of unhashing without `adv_pipe`. I am planning to go back to this and remove these redundancies.
Overall, I am not very convinced with the way I am using memory and pointers and I am sure that there should be a better way to do it.